File: frontend/js/productos.js
let currentUser = null;

async function init() {
  currentUser = await requireSession();
  if (!currentUser) return;
  setUserChip(currentUser);
  updateSidebar(currentUser);

  const path = window.location.pathname;

  if (path.includes('list')) await loadProducts();
  else if (path.includes('create')) initCreateForm();
  else if (path.includes('edit')) {
    const id = new URLSearchParams(window.location.search).get('id');
    if (id) await loadEditForm(id);
  } else if (path.includes('pendiente')) {
    if (currentUser.rol !== 'admin') {
      document.getElementById('content').innerHTML =
        '<div class="alert alert-error">Acceso restringido a administradores.</div>';
      return;
    }
    await loadPending();
  }
}

function updateSidebar(user) {
  const navAdmin = document.getElementById('nav-admin-section');
  const navUsers = document.getElementById('nav-users');
  const navPending = document.getElementById('nav-pending');
  const navCreate = document.getElementById('nav-create');
  const btnCreate = document.getElementById('btn-create');

  if (user.rol === 'admin') {
    if (navAdmin) navAdmin.style.display = 'block';
    if (navUsers) navUsers.style.display = 'flex';
    if (navPending) navPending.style.display = 'flex';
  }
  if (user.rol !== 'demandante') {
    if (navCreate) navCreate.style.display = 'flex';
    if (btnCreate) btnCreate.style.display = 'inline-flex';
  }
}

async function loadProducts() {
  const tbody = document.getElementById('products-tbody');
  tbody.innerHTML = '<tr><td colspan="6" class="loading">Cargando...</td></tr>';

  try {
    const products = await api.get('/productos');
    if (!products.length) {
      tbody.innerHTML = `<tr><td colspan="6">
        <div class="empty-state"><div class="icon">📦</div><h3>Sin productos</h3></div>
      </td></tr>`;
      return;
    }

    tbody.innerHTML = products.map(p => `
      <tr>
        <td><strong>${escHtml(p.title)}</strong></td>
        <td>${escHtml(p.category)}</td>
        <td>Bs. ${parseFloat(p.price).toFixed(2)}</td>
        <td>${getStatusBadge(p.status)}</td>
        <td>${escHtml(p.ofertante_name || '—')}</td>
        <td>
          <div class="actions">
            ${canEdit(p) ? `<a href="/productos/edit.html?id=${p.id}" class="btn btn-sm btn-secondary">Editar</a>` : ''}
            ${canDelete(p) ? `<button class="btn btn-sm btn-danger" data-action="delete-product" data-id="${p.id}" data-title="${escHtml(p.title)}">Eliminar</button>` : ''}
          </div>
        </td>
      </tr>
    `).join('');
  } catch (err) {
    tbody.innerHTML = `<tr><td colspan="6"><div class="alert alert-error">${escHtml(err.message)}</div></td></tr>`;
  }
}

function canEdit(p) {
  return currentUser.rol === 'admin' || (currentUser.rol === 'ofertante' && p.ofertante_id === currentUser.id);
}

function canDelete(p) {
  return currentUser.rol === 'admin' || (currentUser.rol === 'ofertante' && p.ofertante_id === currentUser.id);
}

async function deleteProduct(id, title) {
  if (!confirm(`¿Eliminar el producto "${title}"?`)) return;
  try {
    await api.delete(`/productos/${id}`);
    await loadProducts();
    showAlert('#alert-global', 'Producto eliminado.', 'success');
  } catch (err) {
    showAlert('#alert-global', err.message, 'error');
  }
}

function initCreateForm() {
  if (currentUser.rol === 'demandante') {
    document.getElementById('content').innerHTML =
      '<div class="alert alert-error">Solo ofertantes pueden crear productos.</div>';
    return;
  }

  const form = document.getElementById('product-form');
  form.addEventListener('submit', async (e) => {
    e.preventDefault();
    const data = {
      title: document.getElementById('title').value.trim(),
      description: document.getElementById('description').value.trim(),
      price: parseFloat(document.getElementById('price').value),
      category: document.getElementById('category').value.trim()
    };

    if (!data.title || !data.description || isNaN(data.price) || data.price < 0 || !data.category) {
      showAlert('#alert-box', 'Todos los campos son obligatorios y el precio debe ser ≥ 0.', 'error');
      return;
    }

    try {
      await api.post('/productos', data);
      window.location.href = '/productos/list.html';
    } catch (err) {
      showAlert('#alert-box', err.message, 'error');
    }
  });
}

async function loadEditForm(id) {
  try {
    const product = await api.get(`/productos/${id}`);
    document.getElementById('title').value = product.title;
    document.getElementById('description').value = product.description;
    document.getElementById('price').value = product.price;
    document.getElementById('category').value = product.category;

    const form = document.getElementById('product-form');
    form.addEventListener('submit', async (e) => {
      e.preventDefault();
      const data = {
        title: document.getElementById('title').value.trim(),
        description: document.getElementById('description').value.trim(),
        price: parseFloat(document.getElementById('price').value),
        category: document.getElementById('category').value.trim()
      };

      if (!data.title || !data.description || isNaN(data.price) || data.price < 0 || !data.category) {
        showAlert('#alert-box', 'Todos los campos son obligatorios y el precio debe ser ≥ 0.', 'error');
        return;
      }

      try {
        await api.put(`/productos/${id}`, data);
        window.location.href = '/productos/list.html';
      } catch (err) {
        showAlert('#alert-box', err.message, 'error');
      }
    });
  } catch (err) {
    showAlert('#alert-box', err.message, 'error');
  }
}

async function loadPending() {
  const tbody = document.getElementById('pending-tbody');
  tbody.innerHTML = '<tr><td colspan="6" class="loading">Cargando...</td></tr>';

  try {
    const products = await api.get('/productos/pendientes');
    if (!products.length) {
      tbody.innerHTML = `<tr><td colspan="6">
        <div class="empty-state"><div class="icon">✅</div><h3>Sin productos pendientes</h3></div>
      </td></tr>`;
      return;
    }

    tbody.innerHTML = products.map(p => `
      <tr>
        <td><strong>${escHtml(p.title)}</strong></td>
        <td>${escHtml(p.category)}</td>
        <td>Bs. ${parseFloat(p.price).toFixed(2)}</td>
        <td>${escHtml(p.ofertante_name || '—')}</td>
        <td>${formatDate(p.created_at)}</td>
        <td>
          <div class="actions">
            <button class="btn btn-sm btn-success" data-action="status" data-id="${p.id}" data-status="aprobado">Aprobar</button>
            <button class="btn btn-sm btn-danger" data-action="status" data-id="${p.id}" data-status="rechazado">Rechazar</button>
          </div>
        </td>
      </tr>
    `).join('');
  } catch (err) {
    tbody.innerHTML = `<tr><td colspan="6"><div class="alert alert-error">${escHtml(err.message)}</div></td></tr>`;
  }
}

async function changeStatus(id, status) {
  try {
    const res = await api.patch(`/productos/${id}/status`, { status });
    showAlert('#alert-global', res.message || `Producto ${status}.`, 'success');
    await loadPending();
  } catch (err) {
    showAlert('#alert-global', err.message, 'error');
  }
}

document.addEventListener('click', function (e) {
  const btn = e.target.closest('[data-action]');
  if (!btn) return;
  const action = btn.dataset.action;
  const id = Number(btn.dataset.id);
  if (action === 'delete-product') deleteProduct(id, btn.dataset.title);
  if (action === 'status') changeStatus(id, btn.dataset.status);
});

document.addEventListener('DOMContentLoaded', init);

File: app/routes/productos.py
from flask import Blueprint, request, session, jsonify
from app.database import get_db
from app.middleware import require_auth, require_admin, require_ofertante

productos_bp = Blueprint('productos', __name__)

PRODUCT_QUERY = '''
    SELECT p.*, u.name AS ofertante_name
    FROM products p
    LEFT JOIN users u ON p.ofertante_id = u.id
'''


def _valid_price(value):
    return isinstance(value, (int, float)) and not isinstance(value, bool) and value >= 0


@productos_bp.route('', methods=['GET'])
@require_auth
def list_products():
    user = session['user']
    conn = get_db()

    if user['rol'] == 'admin':
        products = conn.execute(PRODUCT_QUERY).fetchall()
    elif user['rol'] == 'ofertante':
        products = conn.execute(
            PRODUCT_QUERY + ' WHERE p.ofertante_id = ?', (user['id'],)
        ).fetchall()
    else:
        products = conn.execute(
            PRODUCT_QUERY + " WHERE p.status = 'aprobado'"
        ).fetchall()

    conn.close()
    return jsonify([dict(p) for p in products])


@productos_bp.route('/pendientes', methods=['GET'])
@require_admin
def list_pending():
    conn = get_db()
    products = conn.execute(
        PRODUCT_QUERY + " WHERE p.status = 'pendiente'"
    ).fetchall()
    conn.close()
    return jsonify([dict(p) for p in products])


@productos_bp.route('/<int:id>', methods=['GET'])
@require_auth
def get_product(id):
    user = session['user']
    conn = get_db()
    product = conn.execute(PRODUCT_QUERY + ' WHERE p.id = ?', (id,)).fetchone()
    conn.close()

    if not product:
        return jsonify({'error': 'Producto no encontrado'}), 404

    p = dict(product)
    if user['rol'] == 'demandante' and p['status'] != 'aprobado':
        return jsonify({'error': 'Acceso denegado'}), 403
    if user['rol'] == 'ofertante' and p['ofertante_id'] != user['id']:
        return jsonify({'error': 'Acceso denegado'}), 403

    return jsonify(p)


@productos_bp.route('', methods=['POST'])
@require_ofertante
def create_product():
    user = session['user']
    data = request.get_json() or {}

    title = data.get('title', '').strip()
    description = data.get('description', '').strip()
    price = data.get('price')
    category = data.get('category', '').strip()

    if not all([title, description, category]):
        return jsonify({'error': 'Todos los campos son obligatorios'}), 400
    if not _valid_price(price):
        return jsonify({'error': 'El precio debe ser un número mayor o igual a 0'}), 400

    conn = get_db()
    cursor = conn.execute(
        'INSERT INTO products (title, description, price, category, ofertante_id) VALUES (?, ?, ?, ?, ?)',
        (title, description, price, category, user['id'])
    )
    conn.commit()
    product = conn.execute(PRODUCT_QUERY + ' WHERE p.id = ?', (cursor.lastrowid,)).fetchone()
    conn.close()
    return jsonify(dict(product)), 201


@productos_bp.route('/<int:id>', methods=['PUT'])
@require_auth
def update_product(id):
    user = session['user']
    conn = get_db()
    product = conn.execute('SELECT * FROM products WHERE id = ?', (id,)).fetchone()

    if not product:
        conn.close()
        return jsonify({'error': 'Producto no encontrado'}), 404

    p = dict(product)
    if user['rol'] == 'demandante':
        conn.close()
        return jsonify({'error': 'Acceso denegado'}), 403
    if user['rol'] == 'ofertante' and p['ofertante_id'] != user['id']:
        conn.close()
        return jsonify({'error': 'Acceso denegado'}), 403

    data = request.get_json() or {}
    title = data.get('title', p['title'])
    description = data.get('description', p['description'])
    price = data.get('price', p['price'])
    category = data.get('category', p['category'])

    if not _valid_price(price):
        conn.close()
        return jsonify({'error': 'El precio debe ser un número mayor o igual a 0'}), 400

    conn.execute(
        "UPDATE products SET title=?, description=?, price=?, category=?, status='pendiente', updated_at=CURRENT_TIMESTAMP WHERE id=?",
        (title, description, price, category, id)
    )
    conn.commit()
    updated = conn.execute(PRODUCT_QUERY + ' WHERE p.id = ?', (id,)).fetchone()
    conn.close()
    return jsonify(dict(updated))


@productos_bp.route('/<int:id>', methods=['DELETE'])
@require_auth
def delete_product(id):
    user = session['user']
    conn = get_db()
    product = conn.execute('SELECT * FROM products WHERE id = ?', (id,)).fetchone()

    if not product:
        conn.close()
        return jsonify({'error': 'Producto no encontrado'}), 404

    p = dict(product)
    if user['rol'] == 'demandante':
        conn.close()
        return jsonify({'error': 'Acceso denegado'}), 403
    if user['rol'] == 'ofertante' and p['ofertante_id'] != user['id']:
        conn.close()
        return jsonify({'error': 'Acceso denegado'}), 403

    conn.execute('DELETE FROM products WHERE id = ?', (id,))
    conn.commit()
    conn.close()
    return jsonify({'message': 'Producto eliminado'})


@productos_bp.route('/<int:id>/status', methods=['PATCH'])
@require_admin
def change_status(id):
    data = request.get_json() or {}
    status = data.get('status')

    if status not in ['aprobado', 'rechazado', 'pendiente']:
        return jsonify({'error': 'Estado inválido. Use: aprobado, rechazado o pendiente'}), 400

    conn = get_db()
    product = conn.execute('SELECT id FROM products WHERE id = ?', (id,)).fetchone()
    if not product:
        conn.close()
        return jsonify({'error': 'Producto no encontrado'}), 404

    conn.execute(
        'UPDATE products SET status=?, updated_at=CURRENT_TIMESTAMP WHERE id=?',
        (status, id)
    )
    conn.commit()
    conn.close()
    return jsonify({'message': f'Producto {status} correctamente'})
